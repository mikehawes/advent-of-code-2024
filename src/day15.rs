mod move_robot;
pub mod robot_plan;
mod warehouse;
mod warehouse_box;
