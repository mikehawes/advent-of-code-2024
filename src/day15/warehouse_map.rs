const WALL: char = '#';
const BOX: char = 'O';
const EMPTY: char = '.';
const ROBOT: char = '@';

type Point = [usize; 2];

pub struct WarehouseMap {
    tiles: Vec<Vec<char>>,
    width: usize,
    height: usize,
    robot_position: Point,
}

impl WarehouseMap {
    pub fn parse(string: &str) -> WarehouseMap {
        let tiles: Vec<Vec<char>> = string.lines().map(|line| line.chars().collect()).collect();
        let width = tiles.first().map(|line| line.len()).unwrap_or(0);
        let height = tiles.len();
        let robot_position = find_robot(&tiles, width, height);
        WarehouseMap {
            tiles,
            width,
            height,
            robot_position,
        }
    }
}

fn find_robot(tiles: &[Vec<char>], width: usize, height: usize) -> Point {
    (0..height)
        .flat_map(|y| (0..width).map(move |x| [x, y]))
        .find(|[x, y]| tiles[*y][*x] == ROBOT)
        .unwrap_or([width, height])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn can_parse_map() {
        let string = "\
            ########\n\
            #..O.O.#\n\
            ##@.O..#\n\
            #...O..#\n\
            #.#.O..#\n\
            #...O..#\n\
            #......#\n\
            ########\n";
        let map = WarehouseMap::parse(string);
        assert_eq!(print(&map), string);
        assert_eq!(map.robot_position, [2, 2]);
    }

    fn print(map: &WarehouseMap) -> String {
        let mut string = "".to_string();
        for y in 0..map.height {
            for x in 0..map.width {
                string.push(map.tiles[y][x]);
            }
            string.push('\n');
        }
        string
    }
}
