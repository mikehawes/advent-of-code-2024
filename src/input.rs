use std::fs::read_to_string;
use std::io;
use std::path::{Path, PathBuf};

pub fn input_to_string<P: AsRef<Path>>(path: P) -> io::Result<String> {
    let full_path = file_in_src(path);
    read_to_string(full_path)
}

pub fn file_in_src<P: AsRef<Path>>(path: P) -> PathBuf {
    let mut full_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    full_path.push("src");
    full_path.push(path);
    full_path
}
