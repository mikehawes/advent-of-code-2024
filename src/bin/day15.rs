use advent_of_code_2024::day15::robot_plan::RobotPlan;
use advent_of_code_2024::input::input_to_string;

fn main() {
    let input = input_to_string("day15/input.txt").unwrap();
    let plan = RobotPlan::parse(input.as_str());
    println!("Part 1: {}", plan.sum_gps_coordinates_at_end());
}
