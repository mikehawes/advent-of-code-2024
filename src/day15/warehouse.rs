use crate::day15::move_robot::{move_robot, Direction};
use std::collections::HashMap;

pub const WALL: char = '#';
pub const BOX: char = 'O';
pub const BOX_LEFT: char = '[';
pub const BOX_RIGHT: char = ']';
pub const EMPTY: char = '.';
pub const ROBOT: char = '@';

pub type Point = [usize; 2];

#[derive(Clone)]
pub struct Warehouse {
    width: usize,
    height: usize,
    walls: Vec<Vec<char>>,
    robot_position: Point,
    box_position_to_char: HashMap<Point, char>,
}

impl Warehouse {
    pub fn parse(string: &str) -> Warehouse {
        let map: Vec<Vec<char>> = string.lines().map(|line| line.chars().collect()).collect();
        let width = map.first().map(|line| line.len()).unwrap_or(0);
        let height = map.len();
        let robot_position = find_robot(&map, width, height);
        let box_position_to_char = find_boxes(&map, width, height);
        let walls = map_walls(&map);
        Warehouse {
            width,
            height,
            walls,
            robot_position,
            box_position_to_char,
        }
    }
    pub fn move_robot(&self, direction: Direction) -> Warehouse {
        move_robot(self, direction)
    }
    pub fn robot_position(&self) -> Point {
        self.robot_position
    }
    pub fn sum_gps_coordinates(&self) -> usize {
        (0..self.height)
            .flat_map(|y| (0..self.width).map(move |x| [x, y]))
            .filter(|point| self.contents(*point) == BOX)
            .map(gps_coordinate)
            .sum()
    }
    pub fn scale_up(&self) -> Warehouse {
        let tiles = self.walls.iter().map(scale_up_line).collect();
        let [x, y] = self.robot_position;
        let robot_position = [x * 2, y];
        let box_position_to_char = self
            .box_position_to_char
            .keys()
            .flat_map(|[x, y]| [([*x * 2, *y], BOX_LEFT), ([*x * 2 + 1, *y], BOX_RIGHT)])
            .collect();
        Warehouse {
            walls: tiles,
            width: self.width * 2,
            height: self.height,
            robot_position,
            box_position_to_char: box_position_to_char,
        }
    }
    fn contents(&self, point: Point) -> char {
        if point == self.robot_position {
            ROBOT
        } else {
            self.tile(point)
        }
    }
    pub fn tile(&self, point: Point) -> char {
        let [x, y] = point;
        *self
            .box_position_to_char
            .get(&point)
            .unwrap_or_else(|| &self.walls[y][x])
    }
    pub fn is_on_map(&self, point: Point) -> bool {
        let [x, y] = point;
        x < self.width && y < self.height
    }
    pub fn set_robot(&mut self, position: Point) {
        self.robot_position = position;
    }
    pub fn write_tile(&mut self, point: Point, contents: char) {
        match contents {
            BOX | BOX_LEFT | BOX_RIGHT => {
                self.box_position_to_char.insert(point, contents);
            }
            EMPTY => {
                self.box_position_to_char.remove(&point);
            }
            _ => {}
        }
    }
}

fn find_robot(tiles: &[Vec<char>], width: usize, height: usize) -> Point {
    (0..height)
        .flat_map(|y| (0..width).map(move |x| [x, y]))
        .find(|[x, y]| tiles[*y][*x] == ROBOT)
        .unwrap_or([width, height])
}

fn find_boxes(tiles: &[Vec<char>], width: usize, height: usize) -> HashMap<Point, char> {
    (0..height)
        .flat_map(|y| (0..width).map(move |x| [x, y]))
        .map(|[x, y]| ([x, y], tiles[y][x]))
        .filter(|(_, tile)| [BOX, BOX_LEFT, BOX_RIGHT].contains(tile))
        .collect()
}

fn map_walls(tiles: &[Vec<char>]) -> Vec<Vec<char>> {
    tiles
        .iter()
        .map(|line| {
            line.iter()
                .map(|tile| match *tile {
                    WALL => WALL,
                    _ => EMPTY,
                })
                .collect()
        })
        .collect()
}

fn gps_coordinate(point: Point) -> usize {
    point[1] * 100 + point[0]
}

fn scale_up_line(line: &Vec<char>) -> Vec<char> {
    line.iter()
        .map(|c| match *c {
            WALL => "##",
            _ => "..",
        })
        .flat_map(|str| str.chars())
        .collect()
}

#[cfg(test)]
pub mod tests {
    use super::*;
    use insta::assert_snapshot;

    #[test]
    fn can_parse_map() {
        let string = "\
            ########\n\
            #..O.O.#\n\
            ##@.O..#\n\
            #...O..#\n\
            #.#.O..#\n\
            #...O..#\n\
            #......#\n\
            ########\n";
        let warehouse = Warehouse::parse(string);
        assert_eq!(print(&warehouse), string);
        assert_eq!(warehouse.robot_position, [2, 2]);
    }

    #[test]
    fn can_find_gps_coordinate() {
        assert_eq!(gps_coordinate([4, 1]), 104)
    }

    #[test]
    fn can_sum_gps_coordinates() {
        let string = "\
            #######\n\
            #...O.O\n\
            #..@...\n";
        let warehouse = Warehouse::parse(string);
        assert_eq!(warehouse.sum_gps_coordinates(), 210)
    }

    #[test]
    fn can_scale_up() {
        let string = "\
            ########\n\
            #..O.O.#\n\
            ##@.O..#\n\
            #...O..#\n\
            #.#.O..#\n\
            #...O..#\n\
            #......#\n\
            ########\n";
        let warehouse = Warehouse::parse(string);
        assert_snapshot!(print(&warehouse.scale_up()))
    }

    pub fn print(warehouse: &Warehouse) -> String {
        let mut string = "".to_string();
        for y in 0..warehouse.height {
            for x in 0..warehouse.width {
                string.push(warehouse.contents([x, y]));
            }
            string.push('\n');
        }
        string
    }
}
