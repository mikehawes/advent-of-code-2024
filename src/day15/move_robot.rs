use crate::day15::warehouse::{Point, Warehouse};
use crate::day15::warehouse_box::WarehouseBox;
use std::collections::HashSet;

#[derive(Copy, Clone)]
pub enum Direction {
    Up,
    Right,
    Left,
    Down,
}

pub fn move_robot(warehouse: &Warehouse, direction: Direction) -> Warehouse {
    let new_position = next_point(warehouse.robot_position(), direction);
    let mut new_warehouse = warehouse.clone();
    if !warehouse.is_on_map(new_position) || warehouse.is_wall(new_position) {
        return new_warehouse;
    }
    match warehouse.box_at(new_position) {
        Some(b) => {
            if move_boxes(warehouse, &mut new_warehouse, b, direction) {
                new_warehouse.set_robot(new_position);
            }
        }
        None => new_warehouse.set_robot(new_position),
    }
    new_warehouse
}

fn move_boxes(
    warehouse: &Warehouse,
    new_warehouse: &mut Warehouse,
    first_box: &WarehouseBox,
    direction: Direction,
) -> bool {
    let mut boxes = vec![first_box];
    let mut to_move = HashSet::from([first_box.number()]);
    loop {
        let points: Vec<Point> = boxes
            .iter()
            .flat_map(|b| b.points_in_dir(direction))
            .map(|p| next_point(p, direction))
            .collect();
        boxes = vec![];
        for point in points {
            if let Some(b) = warehouse.box_at(point) {
                boxes.push(b);
            } else if !warehouse.is_on_map(point) || warehouse.is_wall(point) {
                return false;
            }
        }
        if boxes.is_empty() {
            new_warehouse.move_boxes(&to_move, direction);
            return true;
        }
        boxes.iter().map(|b| b.number()).for_each(|number| {
            to_move.insert(number);
        });
    }
}

fn next_point(point: Point, direction: Direction) -> Point {
    let [x, y] = point;
    match direction {
        Direction::Up => [x, y.wrapping_sub(1)],
        Direction::Down => [x, y + 1],
        Direction::Left => [x.wrapping_sub(1), y],
        Direction::Right => [x + 1, y],
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::day15::move_robot::Direction::{Down, Left, Right, Up};
    use crate::day15::warehouse::tests::print;

    #[test]
    fn can_move_robot() {
        let string = "\
            ...\n\
            .@.\n\
            ...\n";
        let before = Warehouse::parse(string);
        let after = [Up, Down, Left, Right].map(|d| print(&before.move_robot(d)));
        assert_eq!(
            after,
            [
                ".@.\n\
                 ...\n\
                 ...\n",
                "...\n\
                 ...\n\
                 .@.\n",
                "...\n\
                 @..\n\
                 ...\n",
                "...\n\
                 ..@\n\
                 ...\n"
            ]
        )
    }

    #[test]
    fn can_push_boxes() {
        let before = Warehouse::parse("@OO.");
        let after = before.move_robot(Right);
        assert_eq!(print(&after), ".@OO\n")
    }

    #[test]
    fn can_push_big_boxes_right() {
        let before = Warehouse::parse("@[][].");
        let after = before.move_robot(Right);
        assert_eq!(print(&after), ".@[][]\n")
    }

    #[test]
    fn can_push_big_boxes_left() {
        let before = Warehouse::parse(".[][]@");
        let after = before.move_robot(Left);
        assert_eq!(print(&after), "[][]@.\n")
    }

    #[test]
    fn can_push_big_boxes_down() {
        let string = "\
            .@..\n\
            .[].\n\
            []..\n\
            ..[]";
        let before = Warehouse::parse(string);
        let after = before.move_robot(Down);
        let expected = "\
            ....\n\
            .@..\n\
            .[].\n\
            [][]\n";
        assert_eq!(print(&after), expected)
    }

    #[test]
    fn can_push_chain_of_boxes() {
        let string = "\
            @...\n\
            []..\n\
            .[].\n\
            ..[]\n\
            ....";
        let before = Warehouse::parse(string);
        let after = before.move_robot(Down);
        let expected = "\
            ....\n\
            @...\n\
            []..\n\
            .[].\n\
            ..[]\n";
        assert_eq!(print(&after), expected)
    }

    #[test]
    fn can_stop_at_wall() {
        let before = Warehouse::parse("@#");
        let after = before.move_robot(Right);
        assert_eq!(print(&after), "@#\n")
    }

    #[test]
    fn can_stop_at_edge() {
        let before = Warehouse::parse("@");
        let after = before.move_robot(Right);
        assert_eq!(print(&after), "@\n")
    }

    #[test]
    fn can_stop_pushing_at_wall() {
        let before = Warehouse::parse("@O#");
        let after = before.move_robot(Right);
        assert_eq!(print(&after), "@O#\n")
    }

    #[test]
    fn can_stop_pushing_at_edge() {
        let before = Warehouse::parse("@O");
        let after = before.move_robot(Right);
        assert_eq!(print(&after), "@O\n")
    }
}
