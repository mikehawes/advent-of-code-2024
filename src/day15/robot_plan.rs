use crate::day15::move_robot::Direction;
use crate::day15::move_robot::Direction::{Down, Left, Right, Up};
use crate::day15::warehouse::Warehouse;

pub struct RobotPlan {
    warehouse: Warehouse,
    directions: Vec<Direction>,
}

impl RobotPlan {
    pub fn parse(string: &str) -> RobotPlan {
        let parts: Vec<&str> = string.split("\n\n").collect();
        let warehouse = Warehouse::parse(parts[0]);
        let directions = parse_directions(parts[1]);
        RobotPlan {
            warehouse,
            directions,
        }
    }
    pub fn sum_gps_coordinates_at_end(&self) -> usize {
        self.follow().sum_gps_coordinates()
    }
    pub fn scale_up(&self) -> RobotPlan {
        RobotPlan {
            warehouse: self.warehouse.scale_up(),
            directions: self.directions.clone(),
        }
    }
    fn follow(&self) -> Warehouse {
        let mut warehouse = self.warehouse.clone();
        for direction in self.directions.iter() {
            warehouse = warehouse.move_robot(*direction);
        }
        warehouse
    }
}

fn parse_directions(string: &str) -> Vec<Direction> {
    string
        .lines()
        .flat_map(|line| line.chars())
        .flat_map(to_direction)
        .collect()
}

fn to_direction(c: char) -> Option<Direction> {
    match c {
        '^' => Some(Up),
        '>' => Some(Right),
        'v' => Some(Down),
        '<' => Some(Left),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::day15::warehouse;
    use crate::input::input_to_string;
    use insta::assert_snapshot;

    #[test]
    fn can_find_start() {
        let string = input_to_string("day15/small-example.txt").unwrap();
        let plan = RobotPlan::parse(&string);
        assert_snapshot!(print_at_start(&plan))
    }

    #[test]
    fn can_follow_plan() {
        let string = input_to_string("day15/small-example.txt").unwrap();
        let plan = RobotPlan::parse(&string);
        assert_snapshot!(print_warehouse(&plan.follow()))
    }

    #[test]
    fn can_follow_larger_plan() {
        let string = input_to_string("day15/example.txt").unwrap();
        let plan = RobotPlan::parse(&string);
        assert_snapshot!(print_warehouse(&plan.follow()))
    }

    #[test]
    fn can_follow_scaled_up_larger_plan() {
        let string = input_to_string("day15/example.txt").unwrap();
        let plan = RobotPlan::parse(&string);
        assert_snapshot!(print_warehouse(&plan.scale_up().follow()))
    }

    fn print_at_start(plan: &RobotPlan) -> String {
        warehouse::tests::print(&plan.warehouse)
    }

    fn print_warehouse(warehouse: &Warehouse) -> String {
        warehouse::tests::print(warehouse)
    }
}
