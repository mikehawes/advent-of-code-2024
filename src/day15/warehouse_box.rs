use crate::day15::move_robot::Direction;
use crate::day15::warehouse::{Point, BOX_LEFT, BOX_RIGHT};

#[derive(Clone)]
pub struct WarehouseBox {
    number: usize,
    position: Point,
    width: usize,
}

impl WarehouseBox {
    pub fn from_tile_at_point(number: usize, tile: char, point: Point) -> WarehouseBox {
        let [x, y] = point;
        match tile {
            BOX_LEFT => WarehouseBox {
                number,
                position: point,
                width: 2,
            },
            BOX_RIGHT => WarehouseBox {
                number,
                position: [x - 1, y],
                width: 2,
            },
            _ => WarehouseBox {
                number,
                position: point,
                width: 1,
            },
        }
    }
    pub fn points_in_dir(&self, direction: Direction) -> Vec<Point> {
        let [x, y] = self.position;
        match direction {
            Direction::Left => vec![[x, y]],
            Direction::Right => vec![[x + self.width - 1, y]],
            _ => (0..self.width).map(move |i| [x + i, y]).collect(),
        }
    }
    pub fn points_iter(&self) -> impl Iterator<Item = Point> {
        let [x, y] = self.position;
        (0..self.width).map(move |i| [x + i, y])
    }
    pub fn number(&self) -> usize {
        self.number
    }
    pub fn scale_up(&self) -> WarehouseBox {
        let [x, y] = self.position;
        WarehouseBox {
            number: self.number,
            position: [x * 2, y],
            width: self.width * 2,
        }
    }
    pub fn char_at(&self, point: Point) -> char {
        if self.width == 1 {
            return 'O';
        }
        let offset = point[0] - self.position[0];
        if offset == 0 {
            '['
        } else {
            ']'
        }
    }
    pub fn move_dir(&self, direction: Direction) -> WarehouseBox {
        let [x, y] = self.position;
        let position = match direction {
            Direction::Up => [x, y - 1],
            Direction::Down => [x, y + 1],
            Direction::Left => [x - 1, y],
            Direction::Right => [x + 1, y],
        };
        WarehouseBox {
            number: self.number,
            position,
            width: self.width,
        }
    }
}
