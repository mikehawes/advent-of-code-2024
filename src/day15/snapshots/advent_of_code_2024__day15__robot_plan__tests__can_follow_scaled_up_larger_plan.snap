---
source: src/day15/robot_plan.rs
assertion_line: 88
expression: print_warehouse(&plan.scale_up().follow())
snapshot_kind: text
---
####################
##[].......[].[][]##
##[]...........[].##
##[]........[][][]##
##[]......[]....[]##
##..##......[]....##
##..[]............##
##..@......[].[][]##
##......[][]..[]..##
####################
